# Atividade 1 — Exercícios com os comandos básicos de pandas para EDA


Responda as perguntas abaixo no notebook (cada resposta deve ter ao menos uma célula de código que produza o output e, logo abaixo, uma célula de texto com a sua interpretação).

## 1) Carregue o dataset e mostre as primeiras 10 linhas.

In [6]:
import pandas as pd
df = pd.read_csv('/content/emendas.csv')
display(df.head(10))

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"700000,00","0,00","0,00","0,00","690900,00","9100,00"
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"487500,00","0,00","0,00","0,00","487500,00","0,00"
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"100000,00","0,00","0,00","0,00","0,00","100000,00"
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","243750,00","0,00"
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"1658333,94","0,00","0,00","0,00","527200,00","1131133,94"
5,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPINAS DO SUL - RS,4303806,CAMPINAS DO SUL,4300000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","0,00","243750,00"
6,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,MORENO - PE,2609402,MORENO,2600000,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0000,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,"700000,00","0,00","0,00","0,00","700000,00","0,00"
7,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,NOVA FLORESTA - PB,2510105,NOVA FLORESTA,2500000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"278700,00","0,00","0,00","0,00","0,00","278700,00"
8,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTIAGO DO SUL - SC,4215695,SANTIAGO DO SUL,4200000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"300000,00","300000,00","300000,00","0,00","0,00","0,00"
9,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CIANORTE - PR,4105508,CIANORTE,4100000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"245850,00","0,00","0,00","0,00","0,00","245850,00"


Explique em uma frase o que você observa nas primeiras 10 linhas.

As primeiras 10 linhas (0-9) mostram registros de emendas parlamentares individuais do ano de 2014, todas do tipo "Transferências com Finalidade", destinadas a municípios de diferentes estados brasileiros, mas com vários campos de identificação preenchidos como "Sem Informação", o que indica bastante inconsistência nos dados

## 2) Quantas linhas e colunas o dataset possui?

In [22]:
df.shape

(20632, 28)

Explique o que esses números significam.

significa que tem 20632 linhas e 28 colunas

## 3) Liste os nomes das colunas e descreva brevemente o que você entende que cada coluna representa.

In [24]:

df.columns.tolist()


['Código da Emenda',
 'Ano da Emenda',
 'Tipo de Emenda',
 'Código do Autor da Emenda',
 'Nome do Autor da Emenda',
 'Número da emenda',
 'Localidade de aplicação do recurso',
 'Código Município IBGE',
 'Município',
 'Código UF IBGE',
 'UF',
 'Região',
 'Código Função',
 'Nome Função',
 'Código Subfunção',
 'Nome Subfunção',
 'Código Programa',
 'Nome Programa',
 'Código Ação',
 'Nome Ação',
 'Código Plano Orçamentário',
 'Nome Plano Orçamentário',
 'Valor Empenhado',
 'Valor Liquidado',
 'Valor Pago',
 'Valor Restos A Pagar Inscritos',
 'Valor Restos A Pagar Cancelados',
 'Valor Restos A Pagar Pagos']

Escreva sua breve descrição das colunas, como se estivesse criando um dicionário de variáveis para que leitor entenda.





### Dicionário de Variáveis

| Coluna | Descrição |
|---|---|
| `Código da Emenda` | Identificador único da emenda parlamentar |
| `Ano da Emenda` | Ano em que a emenda foi registrada |
| `Tipo de Emenda` | Classificação da emenda (ex: individual, bancada) |
| `Código do Autor da Emenda` | Identificador do parlamentar autor da emenda |
| `Nome do Autor da Emenda` | Nome do parlamentar que propôs a emenda |
| `Número da Emenda` | Número sequencial da emenda |
| `Localidade de aplicação do recurso` | Local onde o recurso será aplicado |
| `Código Município IBGE` | Código oficial IBGE do município beneficiado |
| `Município` | Nome do município beneficiado |
| `Código UF IBGE` | Código numérico IBGE do estado |
| `UF` | Sigla do estado beneficiado |
| `Região` | Região do Brasil onde o recurso é aplicado |
| `Código Função` | Código da função orçamentária (ex: saúde, educação) |
| `Nome Função` | Nome da função orçamentária |
| `Código Subfunção` | Código da subfunção orçamentária |
| `Nome Subfunção` | Detalhamento da subfunção orçamentária |
| `Código Programa` | Código do programa governamental vinculado |
| `Nome Programa` | Nome do programa governamental vinculado |
| `Código Ação` | Código da ação orçamentária específica |
| `Nome Ação` | Descrição da ação orçamentária |
| `Código Plano Orçamentário` | Código do plano orçamentário |
| `Nome Plano Orçamentário` | Descrição do plano orçamentário |
| `Valor Empenhado` | Valor comprometido/reservado para a emenda |
| `Valor Liquidado` | Valor confirmado após prestação de contas |
| `Valor Pago` | Valor efetivamente pago ao beneficiário |
| `Valor Restos A Pagar Inscritos` | Valor inscrito em restos a pagar (não pago no ano) |
| `Valor Restos A Pagar Cancelados` | Valor de restos a pagar que foi cancelado |
| `Valor Restos A Pagar Pagos` | Valor de restos a pagar que foi efetivamente





## 4) Existem valores nulos? quantos?


In [25]:

df.isnull().sum()

,0
Código da Emenda,0
Ano da Emenda,0
Tipo de Emenda,0
Código do Autor da Emenda,0
Nome do Autor da Emenda,0
Número da emenda,0
Localidade de aplicação do recurso,0
Código Município IBGE,0
Município,0
Código UF IBGE,0


sim exite nulos e sao no total 12


| Coluna | Nulos |
|---|---|
| `Código Programa` | 1 |
| `Nome Programa` | 1 |
| `Código Ação` | 1 |
| `Nome Ação` | 1 |
| `Código Plano Orçamentário` | 1 |
| `Nome Plano Orçamentário` | 1 |
| `Valor Empenhado` | 1 |
| `Valor Liquidado` | 1 |
| `Valor Pago` | 1 |
| `Valor Restos A Pagar Inscritos` | 1 |
| `Valor Restos A Pagar Cancelados` | 1 |
| `Valor Restos A Pagar Pagos` | 1 |



## 5) Mostre os tipos das variáveis (se int, string, boolean, float...) e analise se alguma variável parece ter o tipo errado. Faça isso da maneira mais objetiva e otimizada possível.

In [23]:
df.dtypes

,0
Código da Emenda,object
Ano da Emenda,int64
Tipo de Emenda,object
Código do Autor da Emenda,object
Nome do Autor da Emenda,object
Número da emenda,object
Localidade de aplicação do recurso,object
Código Município IBGE,object
Município,object
Código UF IBGE,int64


Se alguma coluna estiver com o tipo errado, escreva abaixo qual é e por que isso pode atrapalhar análises.

**Dica:** Se alguma variável tiver no tipo o nome "object", isso se refere a texto.


As colunas de valores financeiros estão como texto (object) em vez de float64, o que impede qualquer cálculo matemático sobre elas. Alem de que o python interpreta como string


## 6) Usando a coluna `Ano da Emenda` como base, descubra até que ano o dataset está atualizado (mostre as 8 primeiras linhas do df ordenado desc).

In [13]:
df_sorted = df.sort_values('Ano da Emenda', ascending=False)
display(df_sorted.head(8))
print(df['Ano da Emenda'].max())

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
20623,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,SÃO JOÃO BATISTA DO GLÓRIA - MG,3162203,SÃO JOÃO BATISTA DO GLÓRIA,3100000,...,0E53,AQUISICAO DE VEICULOS PARA O TRANSPORTE ESCOLA...,ECOM,EMENDA DE COMISSAO,"2339,00","0,00","0,00","0,00","0,00","2339,00"
20622,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,SERRA - ES,3205002,SERRA,3200000,...,7S51,CONSTRUCAO DE CONTORNO RODOVIARIO (CONTORNO DE...,ECOM,EMENDA DE COMISSAO,"441463,00","0,00","0,00","0,00","0,00","441463,00"
20621,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,VIADUTOS - RS,4322905,VIADUTOS,4300000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,ECOM,EMENDA DE COMISSAO,"292500,00","0,00","0,00","0,00","0,00","292500,00"
20620,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,SANTA ROSA - RS,4317202,SANTA ROSA,4300000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,ECOM,EMENDA DE COMISSAO,"100000,00","0,00","0,00","0,00","100000,00","0,00"
20619,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,AGRICOLÂNDIA - PI,2200103,AGRICOLÂNDIA,2200000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,ECOM,EMENDA DE COMISSAO,"679500,00","0,00","0,00","0,00","0,00","679500,00"
20618,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,ITUBERÁ - BA,2917300,ITUBERÁ,2900000,...,7656,"IMPLANTACAO, AMPLIACAO OU MELHORIA DE ACOES E ...",ECOM,EMENDA DE COMISSAO,"500000,00","0,00","0,00","0,00","500000,00","0,00"
20617,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,CABO DE SANTO AGOSTINHO - PE,2602902,CABO DE SANTO AGOSTINHO,2600000,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,ECOM,EMENDA DE COMISSAO,"242100,00","0,00","0,00","0,00","0,00","242100,00"
20599,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,BRASÍLIA DE MINAS - MG,3108602,BRASÍLIA DE MINAS,3100000,...,0E53,AQUISICAO DE VEICULOS PARA O TRANSPORTE ESCOLA...,ECOM,EMENDA DE COMISSAO,"242100,00","0,00","0,00","0,00","0,00","242100,00"


2016


Explique: qual é o ano mais recente presente no dataset?

2016

## 7) execute um  `describe()` e explique uma conclusão simples que você consegue tirar a partir das estatísticas (se algumas colunas deveriam aparecer no describe mas não aparecem, indique o motivo que você acredita que esteja causando esse problema).

In [16]:
display(df.describe())
print(df.dtypes)

,Ano da Emenda,Código UF IBGE
count,20632.000000,2.063200e+04
mean,2014.738755,3.165636e+06
std,0.867050,1.084096e+06
min,2014.000000,-1.400000e+01
25%,2014.000000,2.500000e+06
50%,2014.000000,3.100000e+06
75%,2016.000000,4.100000e+06
max,2016.000000,5.300000e+06


Código da Emenda                      object
Ano da Emenda                          int64
Tipo de Emenda                        object
Código do Autor da Emenda             object
Nome do Autor da Emenda               object
Número da emenda                      object
Localidade de aplicação do recurso    object
Código Município IBGE                 object
Município                             object
Código UF IBGE                         int64
UF                                    object
Região                                object
Código Função                         object
Nome Função                           object
Código Subfunção                      object
Nome Subfunção                        object
Código Programa                       object
Nome Programa                         object
Código Ação                           object
Nome Ação                             object
Código Plano Orçamentário             object
Nome Plano Orçamentário               object
Valor Empe

## 8) Mostre as últimas 5 linhas do dataset e escreva se a última linha parece completa (sem muitos nulos) ou não.

In [17]:
df.tail(5)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
20627,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,CERRO GRANDE DO SUL - RS,4305173,CERRO GRANDE DO SUL,4300000,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,ECOM,EMENDA DE COMISSAO,"242100,00","242100,00","242100,00","0,00","0,00","0,00"
20628,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,JI-PARANÁ - RO,1100122,JI-PARANÁ,1100000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,ECOM,EMENDA DE COMISSAO,"6921600,00","0,00","0,00","0,00","0,00","6921600,00"
20629,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,BRASÍLIA DE MINAS - MG,3108602,BRASÍLIA DE MINAS,3100000,...,0509,APOIO AO DESENVOLVIMENTO DA EDUCACAO BASICA,ECOM,EMENDA DE COMISSAO,"145771,00","0,00","0,00","0,00","0,00","145771,00"
20630,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,ANGICO - TO,1701051,ANGICO,1700000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,ECOM,EMENDA DE COMISSAO,"292500,00","0,00","0,00","0,00","292500,00","0,00"
20631,Sem informação,2016,Emenda de Comissão,-99,RELATOR GERAL,S/I,RIO NEGRO - PR,4122305,RIO NEGRO,4100000,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Interprete a completude da última linha.

A partir da coluna "Código UF IBGE" a linha foi preenchida como NaN  ou seja  é um campo vazio

## 9) existem valores duplicados no dataframe? Caso positivo, quantos?

In [21]:
df.duplicated().sum()

np.int64(0)